### PR TITLE
SFMC Sink Plugin supporting DataExtract and Tracking Events.

### PR DESCRIPTION
**Pull request by shiva-cirus**
Salesforce Marketing Cloud Sink Plugin with support to retrieve objects of type DataExtract and Tracking Events.

### CODE REVIEW

File: docs/MarketingCloud-batchsource.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+`Bounce Event` - will allow user to fetch Tracking Bounce Events,
+
+`Open Event` - will allow user to fetch Tracking Open Events,
+
+`Unsub Event` - will allow user to fetch Tracking UnSubscribe Events,
+
+`Sent Event` - will allow user to fetch Tracking Sent Events,
+
+`UnSent Event` - will allow user to fetch Tracking UnSent Events.
```
Comment (nitinmotgi) [inline docs/MarketingCloud-batchsource.md]: UnSent is not English it’s programming. Could we define this as normal English words please.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudClient.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    return expression;
+  }
+
+  private <T extends ETSoapObject> List<T> fetchObjectData(ETClient client, Class<T> clazz, ETFilter filter)
+    throws ETSdkException {
+
+    ClassLoader oldCL = Thread.currentThread().getContextClassLoader();
+    try {
+      Thread.currentThread().setContextClassLoader(MarketingCloudClient.class.getClassLoader());
+      //ETResponse<T> etResponse = client.retrieve(clazz, new ETFilter());
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudClient.java]: Comment should be removed

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudInputFormat.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+          tableInfo = getTableMetaData(object, "", client);
+          if (tableInfo != null) {
+            tableInfos.add(tableInfo);
+          }
+        }
+      }
+
+      return tableInfos;
+    } catch (Exception e) {
+      LOG.error("Error in fetchTableInfo()", e);
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudInputFormat.java]: This error message is not good. How should user react to this message. Could we have descriptive error message with additional information about how they can fix the problem.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudInputFormat.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  private static MarketingCloudObjectInfo getTableMetaData(SourceObject object, String dataExtensionKey,
+                                                           MarketingCloudClient client) {
+    try {
+      if (object == SourceObject.DATA_EXTENSION) {
+        return client.fetchDataExtensionSchema(dataExtensionKey);
+      } else {
+        return client.fetchObjectSchema(object);
+      }
+    } catch (Exception e) {
+      LOG.error("Error in getTableMetaData()", e);
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudInputFormat.java]: Similar error message here is non intuitive. Can we please please write English (not programming) with clear description on problem and corrective action user can take.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+      }
+
+      if (!iterator.hasNext()) {
+        return false;
+      }
+
+      row = iterator.next();
+
+      pos++;
+    } catch (Exception e) {
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java:89]: Again error message is non intuitive

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+      } else {
+        results = client.fetchObjectRecords(object);
+      }
+
+      LOG.debug("size={}", results.size());
+      if (!results.isEmpty()) {
+        fetchSchema(client);
+      }
+    } catch (Exception e) {
+      results = Collections.emptyList();
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java]: How should user be reacting to this error message.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    Schema.Type fieldType = fieldSchema.getType();
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      return transformLogicalType(fieldName, logicalType, fieldValue);
+    }
+
+    switch (fieldType) {
+      case STRING:
+        return convertToStringValue(fieldValue);
+      case DOUBLE:
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java:236]: Is float not handled ?

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+   * Converts raw field value according to the schema field type.
+   */
+  private Object convertToValue(String fieldName, Schema fieldSchema, Object fieldValue) {
+    Schema.Type fieldType = fieldSchema.getType();
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      return transformLogicalType(fieldName, logicalType, fieldValue);
+    }
+
+    switch (fieldType) {
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java:233]: Seems like not all types are handled. If the types are not supported by salesforce then a clear documentation on what types are supported and how they are mapped should be documented.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link BatchSource} that reads data from multiple objects in Salesforce.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(MarketingCloudConstants.PLUGIN_NAME)
+@Description("Reads from multiple objects in SFMC. " +
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java]: This is a highly technical information. Could we explain this in layperson terms ?

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    collector.getOrThrowException();
+
+    //Get Schema
+    stageConfigurer.setOutputSchema(getSchema(conf.getQueryMode()));
+
+    if (pipelineConfigurer.getEngine() == Engine.SPARK) {
+      pipelineConfigurer.setPipelineProperties(Collections.singletonMap("spark.task.maxFailures", "1"));
+    } else if (pipelineConfigurer.getEngine() == Engine.MAPREDUCE) {
+      Map<String, String> properties = new HashMap<>();
+      properties.put("mapreduce.reduce.maxattempts", "1");
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java:87]: Why is reduce limited ?

Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java:87]: Aren’t there constants for this ?

File: src/main/java/io/cdap/plugin/sfmc/source/util/SchemaBuilder.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    Schema schema = createSchema(column);
+    if (schema == null) {
+      return null;
+    }
+
+    return Schema.Type.NULL == schema.getType()
+      ? Schema.Field.of(name, schema)
+      : Schema.Field.of(name, Schema.nullableOf(schema));
+  }
+
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/sfmc/source/util/SchemaBuilder.java:66]: Looks like types are fixed, some documentation above would be great.

File: docs/MarketingCloud-batchsource.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
+# Salesforce Marketing Cloud Source
+
+Description
+-----------
+
+Reads from one or multiple objects withing Salesforce Marketing Cloud depending on the mode values set for this plugin.
```
Comment (avirmani) [inline docs/MarketingCloud-batchsource.md]: within

File: docs/MarketingCloud-batchsource.md (hunk 1 above)
Comment (shiva-cirus) [reply docs/MarketingCloud-batchsource.md]: Done

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudClient.java (hunk 2 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudClient.java]: Done

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudInputFormat.java (hunk 4 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudInputFormat.java]: Added descriptive message

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java (hunk 6 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java]: Resolved.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java (hunk 5 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java:89]: Added descriptive message

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudInputFormat.java (hunk 3 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudInputFormat.java]: Added descriptive message.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java (hunk 7 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java:236]: Float is not one of the datatypes supported by the FuelSDK API. https://developer.salesforce.com/docs/atlas.en-us.noversion.mc-apis.meta/mc-apis/propertytype.htm

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java (hunk 8 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java:233]: Its documented in the design doc.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java (hunk 9 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java]: Simplified.

File: docs/MarketingCloud-batchsource.md (hunk 12 above)
Comment (shiva-cirus) [reply docs/MarketingCloud-batchsource.md]: Done

File: src/main/java/io/cdap/plugin/sfmc/source/util/SchemaBuilder.java (hunk 11 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/util/SchemaBuilder.java:66]: Added documentation

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudClient.java [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+  /**
+   * Fetch records for passed object from Salesforce Marketing Cloud.
+   *
+   * @param dataExtensionKey The data extension key for which data to be fetched from Salesforce Marketing Cloud
+   * @return The list of ETDataExtensionRow representing the records from requested data extension
+   * @throws ETSdkException The FuelSDKException
+   */
+  public List<ETDataExtensionRow> fetchDataExtensionRecords(String dataExtensionKey, String filterStr)
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudClient.java:130]: looks like this loads the entire contents of the data extension into memory. I would expect this to go out of memory for any non-tiny data extension. What size data extensions was this tested on?

File: docs/MarketingCloud-batchsource.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+**Client Secret**: TOAuth2 client secret associated with an installed package in the Salesforce Marketing Cloud.
+
+**Authentication Base URI**: Authentication Base URL associated for the Server-to-Server API integration. 
+For example, `https://instance.auth.marketingcloudapis.com/`
+
+**SOAP API Endpoint**: The SOAP Endpoint URL associated for the Server-to-Server API integration. For example, 
+`https://instance.soap.marketingcloudapis.com/Service.asmx`
+
+**REST API Base URI**: The REST API Base URL associated for the Server-to-Server API integration. For example, 
+`https://instance.rest.marketingcloudapis.com/`
```
Comment (albertshau) [inline docs/MarketingCloud-batchsource.md]: this doc should have a mapping of SFMC data types to CDAP data types, similar to what is in https://github.com/data-integrations/google-cloud/blob/develop/docs/BigQueryTable-batchsink.md#data-type-mappings-from-cdap-to-bigquery

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java (hunk 8 above)
Comment (albertshau) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java:233]: should be documented in the plugin doc

File: src/main/java/io/cdap/plugin/sfmc/source/util/SchemaBuilder.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+  private Schema createSchema(MarketingCloudColumn column) {
+    //FuelSDK supports the following types -
+    // https://developer.salesforce.com/docs/atlas.en-us.noversion.mc-apis.meta/mc-apis/propertytype.htm
+    switch (column.getTypeName().toLowerCase()) {
+      case "decimal":
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/sfmc/source/util/SchemaBuilder.java:73]: decimals in systems usually map to the numeric CDAP type. What is the actual java object that comes back for a decimal type?

File: src/main/java/io/cdap/plugin/sfmc/source/util/SchemaBuilder.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    return Schema.Type.NULL == schema.getType()
+      ? Schema.Field.of(name, schema)
+      : Schema.Field.of(name, Schema.nullableOf(schema));
+  }
+
+  private Schema createSchema(MarketingCloudColumn column) {
+    //FuelSDK supports the following types -
+    // https://developer.salesforce.com/docs/atlas.en-us.noversion.mc-apis.meta/mc-apis/propertytype.htm
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/sfmc/source/util/SchemaBuilder.java]: this doc only mentions 4 types which don't match with the types in this switch case. Is it the right doc?

**Review by albertshau — COMMENTED** [review]
main concern is that it loads everything into memory. Don't see how it can work except on small data sizes.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java (hunk 8 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java:233]: I will update the Documentation.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudClient.java (hunk 13 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudClient.java:130]: On a 1CPU/ 2GB worker,  client successfully retrieved about 1.6M rows from Data extension.

**Comment by albertshau** [comment]
Merging this, though most of it will need to be revisited.

**Pull request merged**